### ISSUE TITLE
Handle leak

### ISSUE THREAD

**Issue by MBerantzino** [Yubico/yubioath-desktop#752]
- Yubico Authenticator version: 5.1.0
- Operating system and version: Windows 10 x64, 21H1 - fully patched
- YubiKey model and version: Not relevant
- Bug description summary: Yubico Authenticator leaks handles, and eventually renders the Task Manager's "Processes" and "Users" tabs inoperable. Reproduceable on 2 different setups.

Steps to reproduce
---

1) Start the Yubico Authenticator
2) Monitor the "Handles" count tab in Task Manager's "Details" tab (you may have to add the column in Task Manager, right click on headers and click "Select columns")
3) The handle count will increase by 2 every second
4) After about 1,5 hours (handle count at 10.300+) the "Processes" tab and "Users" tab will not display anything (make sure to close Task Manager completely and re-open to see the effect)

Expected result
---

Handle count should not increase - and Task Manager should not be affected by running Yubico Authenticator.

Actual results
---

Handle count increases by 2 every second.

After about 1,5 hours (handle count at 10.300+) the "Processes" tab and "Users" tab will not display anything (make sure to close Task Manager completely and re-open to see the effect)

Other info
---

Windows falsely reports that another user is logged on when shutting Windows down. This is a side-effect of the deeper issue with the leaked handles.

Instead of using Task Manager, one could also run the following command in a console:

query user

When the issue happens, this will report an error.

After the issue has occurred - closing the Yubico Authenticator makes Task Manager operable again.

A downgrade to v5.0.5 solves the problem on both tested systems.

**Comment by webshaun** [comment]
This is undoubtedly related to this.  [https://github.com/Yubico/yubioath-desktop/issues/759](https://github.com/Yubico/yubioath-desktop/issues/759)

**Comment by fdennis** [comment]
@MBerantzino I would appreciate it if you could try to help us in verifying if this issue has been resolved or not in a newer version of the authenticator . Begin by making sure you have access to https://github.com/Yubico/yubioath-desktop/actions/runs/1520264211 ? If yes, there should be an artifact at the bottom of the page containing a .zip file. If you download and extract that .zip file you should get a new directory with a .msi and yet another .zip file. There are two ways of testing this newer executable:
1. Run the .msi installer. Then run the installed version. Or:
2. Extract the second .zip. That should give you a new directory containing a "release" folder. If you open the "release" directory you should find a `yubioath-desktop.exe` executable. Try to run that.

**Please note that neither the installer nor the executable have been signed. The artifact that you download from Github has been automatically created by Github actions. I have tried and verified both of the alternatives above.**

Does the same issue appear with this newer version? When I tried on my Windows machine it seems to have been solved but I would like to get this verified by someone else.

Please let me know if you have any questions.

**Comment by MBerantzino** [comment]
> Does the same issue appear with this newer version? When I tried on my Windows machine it seems to have been solved but I would like to get this verified by someone else.

@fdennis Initial test looks like the issue has been fixed. 

I installed the MSI-version, started the app and have tested with multiple yubikeys and multiple inserts/removal of the keys without noticing any significant increase in the handle count (hovers around 485-490). Any increase in the handle count is reduced again as expected.

As I expect the handle count issue to be the origin of the Task Manager and Windows log off symptoms I would expect these to have been solved as well. However as these symptoms only manifests themselves after around 1,5 hours, I will continue to monitor and report back.

**Comment by fdennis** [comment]
@MBerantzino Great, thank you for your help!

**Comment by MBerantzino** [comment]
@fdennis I haven't noticed any issue with the handle count in regards to the Task Manager and Windows log off symptoms.

I would say that the issue has been fixed.

With the above said, I have noticed a slight increase of the handle count when you navigate between the different views in the app - but this is not a big issue. It appears to be the "About"-view that periodically increases the handle count which aren't freed again - but as it is only around 4 handles and you really want to nitpick to say complain about it.

I did today encounter an issue that I haven't had before: Suddenly neither of my 2 keys could be read (correctly) by the app. Sometimes an error message was displayed ("couldn't read the key" or something like that) - sometimes the list of entries was displayed, but I couldn't get the codes. This all occurred in the same session and a close of the app (=minimize to tray) and reopen fixed the issue. I only encountered the problem 1 time - and could be just a fluke where I switched keys too quickly or something like that. Unless others have reported the same issue, I wouldn't worry about it.
